server/src/main/java/org/opensearch/index/store/CompositeDirectory.java

```diff
@@ -0,0 +1,345 @@
+                .filter(file -> !FileType.isBlockFile(file))
+                .collect(Collectors.toUnmodifiableSet());
+            String[] files = new String[localLuceneFiles.size()];
+            localLuceneFiles.toArray(files);
+            Arrays.sort(files);
+            logger.trace("listAll() returns : {}", () -> Arrays.toString(files));
+            return files;
+        } finally {
+            readLock.unlock();
+        }
+    }
+
+    /**
+     * Removes an existing file in the directory.
+     * Throws {@link NoSuchFileException} or {@link FileNotFoundException} in case file is not present locally and in remote as well
+     * @param name the name of an existing file.
+     * @throws IOException in case of I/O error
+     */
+    @Override
+    public void deleteFile(String name) throws IOException {
+        logger.trace("deleteFile() called {}", name);
+        writeLock.lock();
+        try {
+            localDirectory.deleteFile(name);
+            fileCache.remove(localDirectory.getDirectory().resolve(name));
+        } catch (NoSuchFileException | FileNotFoundException e) {
+            logger.trace("File {} not found in local, trying to delete from Remote", name);
+            try {
+                remoteDirectory.deleteFile(name);
+            } finally {
+                writeLock.unlock();
+            }
+        } finally {
+            writeLock.unlock();
+        }
+    }
+
+    /**
+     * Returns the byte length of a file in the directory.
+     * Throws {@link NoSuchFileException} or {@link FileNotFoundException} in case file is not present locally and in remote as well
+     * @param name the name of an existing file.
+     * @throws IOException in case of I/O error
+     */
+    @Override
+    public long fileLength(String name) throws IOException {
+        logger.trace("fileLength() called {}", name);
+        readLock.lock();
+        try {
+            long fileLength;
+            if (Arrays.asList(getRemoteFiles()).contains(name) == false) {
+                fileLength = localDirectory.fileLength(name);
+                logger.trace("fileLength from Local {}", fileLength);
+            } else {
+                fileLength = remoteDirectory.fileLength(name);
+                logger.trace("fileLength from Remote {}", fileLength);
+            }
+            return fileLength;
+        } finally {
+            readLock.unlock();
+        }
+    }
+
+    /**
+     * Creates a new, empty file in the directory and returns an {@link IndexOutput} instance for
+     * appending data to this file.
+     * @param name the name of the file to create.
+     * @throws IOException in case of I/O error
+     */
+    @Override
+    public IndexOutput createOutput(String name, IOContext context) throws IOException {
+        logger.trace("createOutput() called {}", name);
+        writeLock.lock();
+        try {
+            return localDirectory.createOutput(name, context);
```

Comment:
I have a more generic question, what happens to the segments written by a "warm" shard? Are those segments always treated as "Hot"? Do we want to start treating them as warm as well after some point in time. 
Since we are developing primitives for writeable warm, it is possible that most of the data written for a warm shard is always 'HOT'.
I think we should handle that as well.